### PR TITLE
test(gha): optional backend testing on python 3.6

### PR DESCRIPTION
**Pull request by joshuarli**
Bringing back https://github.com/getsentry/sentry/pull/18241 but this time, on GHA to conserve precious travis jobs.

We begin with the most basic backend testing:

- the "postgres" test suite (there are others - snuba, symbolicator, relay-integration)
- no migrations (testing will build the appropriate db stuff on-the-fly)

Once all tests pass, we'll add the other test suites as well as migrations testing.

Also for now, `rb` is bumped to upstream master only for py3 testing.

### CODE REVIEW

**Comment by joshuarli** [comment]
Okay, my rb fork is a bit outdated. https://github.com/joshuarli/rb/commit/1aa36e0feab75f6cd89dc0c1323e6981c94d858c was pinning redis back.

**Comment by joshuarli** [comment]
Need to cut new release for rb soon, a good requirement would be if it passes sentry py2.

**Comment by joshuarli** [comment]
The overhead is about ~2 min per job - 4 shards seems good given total test time is at least 20 minutes with all the failing stuff right now.

**Comment by EvanPurkhiser** [comment]
What's up with the massive amount of stacktraces

**Comment by joshuarli** [comment]
> What's up with the massive amount of stacktraces

Hmm, I thought b4557d5 would do the trick but the problem remains.

**Comment by joshuarli** [comment]
Oh also, should probably have GHA backend testing on py2 first..?

**Comment by joshuarli** [comment]
Ok, flat out disabling internal metrics (which isn't necessary for tests? I'm not entirely sure yet what it is) works. d7654ca6a49641b4d906138464b649ee4c2e425b

**Comment by joshuarli** [comment]
Disabling the internal metrics for tests probably isn't the right approach here, need to figure out why it's specific to GHA.

FTR disabling them doesn't really result in much of a speedup, and there are still a few test codepaths so coverage isn't much of a concern. Example `tests/sentry/tsdb/test_redis.py:56: in test_make_counter_key`.

**Comment by joshuarli** [comment]
Well that is super weird, it's fine now.

File: .github/workflows/backend.yml
Comment (BYK) [inline .github/workflows/backend.yml:1]: I'd add a `test-` prefix somewhere. Not all workflows are tests 😉

File: .github/workflows/backend.yml
Comment (BYK) [inline .github/workflows/backend.yml:72]: Shouldn't this come _before_ pip cache?

File: .github/workflows/backend.yml
Comment (BYK) [inline .github/workflows/backend.yml:90]: I feel like we can skip/optimize steps until here with simply using a Py3-based docker image. What do you think @joshuarli?

File: .travis.yml
Comment (BYK) [inline .travis.yml:134]: Are these changes related?

File: .github/workflows/backend.yml
Comment (joshuarli) [reply .github/workflows/backend.yml:1]: @billyvg your opinion on this? We should be consistent if I did make this change.

File: .travis.yml
Comment (joshuarli) [reply .travis.yml:134]: Yep, sentry init can't be called twice - the 2nd time will error in CI because it wants an interactive y/n to overwrite the config file.

It also makes more sense to explicitly execute it after installing sentry anyways.

Comment (joshuarli) [reply .travis.yml:134]: Previously this was in the Makefile but in GHA we use devservices to spin up services, and so the actual testing step which uses make did init a 2nd time.

File: .github/workflows/backend.yml
Comment (BYK) [reply .github/workflows/backend.yml:1]: FYI there's already a `release.yml` workflow.

File: .github/workflows/backend.yml
Comment (joshuarli) [reply .github/workflows/backend.yml:90]: Can you elaborate on what could be skipped/optimized? I don't see much point, python deps are cached and so this doesn't take much time.

Comment (joshuarli) [reply .github/workflows/backend.yml:90]: And I think the implementation would be complex and likely slower because you're going to be pulling in a whole linux distribution.

File: .github/workflows/backend.yml
Comment (joshuarli) [reply .github/workflows/backend.yml:72]: I'll defer to @billyvg here, I think ideally it'd be the 1st step, even before actions/checkout, but not sure what convention is.

File: .github/workflows/backend.yml
Comment (joshuarli) [reply .github/workflows/backend.yml:1]: Yeah, I'm talking about the others, they should be renamed too.

Comment (billyvg) [reply .github/workflows/backend.yml:1]: I've been trying to group workflows based on the `on` event. Does this need to  run if only javascript files are changed? I'd move it into the `python` workflow and name the job as `test-py3`

File: .github/workflows/backend.yml
Comment (billyvg) [reply .github/workflows/backend.yml:72]: Yeah I think that makes sense

Comment (joshuarli) [reply .github/workflows/backend.yml:72]: Ok, I can submit a PR for this later.

File: .github/workflows/backend.yml
Comment (joshuarli) [reply .github/workflows/backend.yml:1]: No to javascript. And ok, I'll look into that.

**Comment by BYK** [comment]
Do you think we can also report the number of failing tests for Py3 and have a chart or something for the python apocalypse project?

/cc @logicalbomb

**Comment by joshuarli** [comment]
> Do you think we can also report the number of failing tests for Py3 and have a chart or something for the python apocalypse project?
> 
> /cc @logicalbomb

Evan has something like ~50 failed tests now - if the intent is to show progress over time, there isn't really much to show starting from now.

File: .github/workflows/python.yml
Comment (joshuarli) [inline .github/workflows/python.yml:10]: Had to deindent this job.

File: tests/sentry/__init__.py
Comment (billyvg) [inline tests/sentry/__init__.py:3]: 👁️

**Review by billyvg — APPROVED** [review]

File: .github/workflows/python.yml
Comment (billyvg) [inline .github/workflows/python.yml:108]: doesn't py3 imply backend?

Comment (joshuarli) [reply .github/workflows/python.yml:108]: How so? It could be py2 or py3.

Also, still need testing on py2, but it can probably wait. Because I still have to completely enumerate test configurations for py3 (right now it's only a basic one).

**Comment by joshuarli** [comment]
@EvanPurkhiser Feel free to merge this whenever you want it.

**Comment by EvanPurkhiser** [comment]
Going to merge this when it's passing, so I can start seeing what the real % is on master

File: .github/workflows/python.yml
Comment (EvanPurkhiser) [reply .github/workflows/python.yml:108]: He means it could just be `test-py3`

Comment (billyvg) [reply .github/workflows/python.yml:108]: Yeah, I meant: isn't `backend` redundant?

Comment (joshuarli) [reply .github/workflows/python.yml:108]: Oh, lol, yeah you're right.

File: .github/workflows/python.yml
Comment (joshuarli) [inline .github/workflows/python.yml:199]: https://github.com/getsentry/sentry/pull/20418#issuecomment-681194504

**Pull request merged**
